app/src/main/java/dev/itsvic/parceltracker/api/USPSDeliveryService.kt

```diff
@@ -0,0 +1,74 @@
+package dev.itsvic.parceltracker.api
+
+import dev.itsvic.parceltracker.R
+import okhttp3.MediaType
+import okhttp3.MediaType.Companion.toMediaType
+import okhttp3.Request
+import okhttp3.RequestBody
+import okhttp3.RequestBody.Companion.toRequestBody
+import okhttp3.coroutines.executeAsync
+import android.util.Log
+import kotlinx.coroutines.ExperimentalForInheritanceCoroutinesApi
+import org.json.JSONObject
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import retrofit2.Retrofit
+import retrofit2.http.Path
+
+object USPSDeliveryService : DeliveryService {
+    override val nameResource: Int = R.string.service_usps
+    override val acceptsPostCode: Boolean = false
+    override val requiresPostCode: Boolean = false
+
+    override suspend fun getParcel(trackingId: String, postCode: String?): Parcel {
+
+        return Parcel(
+            trackingId,
+            "history",
+            "status",
+            "metadata"
+        )
+    }
+
+    private val retrofit = Retrofit.Builder().baseUrl("https://api.usps.com/").client(api_client)
+        .addConverterFactory(api_factory).build()
+
+    private val service = retrofit.create(API::class.java)
+
+    private interface API {
+        @GET("tracking")
+        suspend fun getStatus(
+            @Path("trackingId") trackingId: String,
+
+        )
+    }
+
+    @OptIn(ExperimentalCoroutinesApi::class)
+    private suspend fun getOauthToken(clientId: String, clientSecret: String): String {
+        // this is all coming from https://developers.usps.com/Oauth
+        val GRANT_TYPE = "authorization_code"
+        val SCOPES = listOf<String>("tracking")
+        val TOKEN_URL = "https://apis.usps.com/oauth2/v3/token"
```

Comment:
you really should just use Retrofit for this, grep for `@Body` in the source for examples